### PR TITLE
adding bandit support

### PR DESCRIPTION
**Pull request by nishils**
This PR adds bandit support for python projects.

Still working on validating the rulesets but looks promising so far.

### CODE REVIEW

File: lib/salus/scanners/bandit.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+require 'bundler'
+require 'salus/scanners/base'
+
+# Bandit scanner to check for Python vulns.
+# https://github.com/presidentbeef/brakeman
```
Comment (jborrey) [inline lib/salus/scanners/bandit.rb]: this link is not correct right?

File: lib/salus/scanners/bandit.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+module Salus::Scanners
+  class Bandit < Base
+    def run
+      Dir.chdir(@repository.path_to_repo) do
+
+        shell_return = run_shell('bandit -r -ii -f json -a file -x tests .')
```
Comment (jborrey) [inline lib/salus/scanners/bandit.rb]: nit: comments for what each flag does (some are not obvious)

File: lib/salus/scanners/bandit.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          )
+          report_stderr(shell_return.stderr)
+        end
+      end
+    end
+
+    def should_run?
+      @repository.requirements_txt_present? || @repository.setup_cfg_present?
+    end
+
```
Comment (jborrey) [inline lib/salus/scanners/bandit.rb]: nit: remove this blank line - rubocop will probably complain

**Comment by jborrey** [comment]
Thanks for opening this @nishils - Python support would be great.

A while ago I remember testing out Bandit on some _hello world_ flask apps and random pythons scripts and it didn't really find much - that might be different today.

Do you have some examples of what it can detect?

**Comment by xolaz** [comment]
tests please!

File: lib/salus/scanners/bandit.rb (hunk 1 above)
Comment (cbtoni) [reply lib/salus/scanners/bandit.rb:5]: @ghbren Can you update this url to point at bandit?

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+    end
+
+    context 'different config options' do
+      let(:repo) { Salus::Repo.new("#{py_dir}/python_project_vulns") }
+
+      it 'aggregate option should work' do
+        repo = Salus::Repo.new("#{py_dir}/python_project_vulns2")
+
+        # aggregate by file
+        config_file = "#{py_dir}/salus_configs/aggregate_file.yaml"
```
Comment (cbtoni) [inline spec/lib/salus/scanners/bandit_spec.rb:113]: Any reason not to skip file parsing and just pass the config options directly to the scanner?

```ruby
scanner = Salus::Scanners::Bandit.new(repository: repo, config: {"aggregate" => "file"})
```

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+        logs = JSON.parse(scanner.report.to_h[:logs])
+        results = logs['results']
+
+        # results ordered by filename
+        expect(results.size).to eq(7)
+        expect(results[0]['filename']).to eq('./main.py')
+        results[1..-1].each { |r| expect(r['filename']).to eq('./main2.py') }
+
+        # aggregate by vuln
```
Comment (cbtoni) [inline spec/lib/salus/scanners/bandit_spec.rb:127]: Please make this one a separate test

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+        expect(results[2]['test_id']).to eq('B301')
+        expect(results[3]['test_id']).to eq('B301')
+        expect(results[4]['test_id']).to eq('B301')
+        expect(results[5]['test_id']).to eq('B105')
+        expect(results[6]['test_id']).to eq('B105')
+      end
+
+      it 'config file option should work' do
+        repo = Salus::Repo.new("#{py_dir}/python_project_vulns")
+        scanner = Salus::Scanners::Bandit.new(repository: repo, config: {})
```
Comment (cbtoni) [inline spec/lib/salus/scanners/bandit_spec.rb:150]: Do we need to run the scanner before the config here?

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+
+        # exclude 1 file
+        config_file = "#{py_dir}/salus_configs/exclude1.yaml"
+        configs = Salus::Config.new([File.read(config_file)]).scanner_configs['Bandit']
+        scanner = Salus::Scanners::Bandit.new(repository: repo, config: configs)
+        scanner.run
+        logs = JSON.parse(scanner.report.to_h[:logs])
+        confidence = logs['metrics']['_totals']
+
+        expect(scanner.report.passed?).to eq(false)
```
Comment (cbtoni) [inline spec/lib/salus/scanners/bandit_spec.rb:332]: How are these confidence metrics related to excluding paths?

File: lib/salus/scanners/bandit.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    def config_options
+      # config options taken from https://pypi.org/project/bandit/
+
+      string_to_flag_map = {
+        'level' => { 'LOW' => 'l', 'MEDIUM' => 'll', 'HIGH' => 'lll',
```
Comment (cbtoni) [inline lib/salus/scanners/bandit.rb:70]: Maybe instead of defining both "LOW" and "low", we can just do a case-insensitive comparison or similar in `build_flag_args_from_string`?

File: lib/salus/scanners/bandit.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                     'low' => 'l', 'medium' => 'll', 'high' => 'lll' },
+        'confidence' => { 'LOW' => 'i', 'MEDIUM' => 'ii', 'HIGH' => 'iii',
+                          'low' => 'i', 'medium' => 'ii', 'high' => 'iii' }
+      }
+      args = build_flag_args_from_string(string_to_flag_map)
+      args.merge!(aggregate: { type: :string, keyword: 'a' },
```
Comment (cbtoni) [inline lib/salus/scanners/bandit.rb:76]: It looks like in other scanners we use regexes to define allowed valid values in these fields. Can you please add some where appropriate?

**Comment by heimdall-asguard** [comment]
Review Error for cbtoni @ 2020-05-18 23:28:30 UTC
User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)

File: lib/salus/scanners/bandit.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
+require 'bundler'
```
[the remark this answers is not in the document]
Comment (ghbren) [reply lib/salus/scanners/bandit.rb]: removed

File: lib/salus/scanners/bandit.rb (hunk 1 above)
Comment (ghbren) [reply lib/salus/scanners/bandit.rb]: fixed

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+        expect(info[:stdout]).not_to be_empty
+        msg = 'Consider possible security implications associated with cPickle module.'
+        expect(logs).to include(msg)
+      end
+    end
+
+    context 'python project with insecure code in a nested folder' do
+      let(:repo) { Salus::Repo.new("#{py_dir}/python_project_with_insecure_code_practices_r") }
+
+      it 'should record failure and record the STDOUT from gosec' do
```
[the remark this answers is not in the document]
Comment (ghbren) [reply spec/lib/salus/scanners/bandit_spec.rb]: fixed

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+        scanner = Salus::Scanners::Bandit.new(repository: repo, config: {})
+
+        expect(repo.requirements_txt_present?).to eq(false)
+        expect(repo.setup_cfg_present?).to eq(false)
+        expect(scanner.should_run?).to eq(false)
+      end
+    end
+
+    context 'requirements.txt present but no setup.cfg' do
+      it 'should return false' do
```
[the remark this answers is not in the document]
Comment (ghbren) [reply spec/lib/salus/scanners/bandit_spec.rb]: yes, fixed

File: spec/lib/salus/scanners/bandit_spec.rb [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+        scanner = Salus::Scanners::Bandit.new(repository: repo, config: {})
+
+        expect(repo.requirements_txt_present?).to eq(true)
+        expect(repo.setup_cfg_present?).to eq(false)
+        expect(scanner.should_run?).to eq(true)
+      end
+    end
+
+    context 'setup.cfg present but no requirements.txt' do
+      it 'should return false' do
```
[the remark this answers is not in the document]
Comment (ghbren) [reply spec/lib/salus/scanners/bandit_spec.rb]: fixed

File: lib/salus/scanners/bandit.rb (hunk 9 above)
Comment (ghbren) [reply lib/salus/scanners/bandit.rb]: I have updated all appropriate config options (aggregate, tests, skip) to use regexes.  
`level` and `confidence` cannot be updated to regexes, because for example
`level: low` in salus.yaml will be converted to command line option `-i` (with build_flag_args_from_string), and `level: medium` will be convert to `-ii`. There is code that checks whether values passed to `level` and `confidence` are valid though.

File: lib/salus/scanners/bandit.rb (hunk 8 above)
Comment (ghbren) [reply lib/salus/scanners/bandit.rb]: Removed lower case keys from these hashes keys. 
User supplied values on `level`/`confidence` are converted to upper case during comparison.

File: spec/lib/salus/scanners/bandit_spec.rb (hunk 7 above)
Comment (ghbren) [reply spec/lib/salus/scanners/bandit_spec.rb:329]: I was checking if number of vulnerabilities found per confidence level decreased as I excluded paths.  I agree this was not ideal, so I updated the code to check for actual file names in the results instead.

**Comment by ghbren** [comment]
> In general, I'd split your specs up a bit, and have one spec test for one thing. Also, I think testing that bandit is receiving one configuration value is helpful, but I'm not sure we need to test every options possible value. (i.e. testing for severity = 'high' is probably going to work similarly to severity = 'low', 'medium', etc.)

I restructured the whole test file to so that each `it` section tests one config option value.

I tested extra possible values to be safe.  For example, even if I don't add use the `severity` option at all, all vulnerabilities found may have `severity=high`. So if my wrapper for severity is not working and I test setting `severity=high`, the results may still look correct.  
The extra values will show differences in severity results, so we are sure the severity wrapper is working.

File: spec/lib/salus/scanners/bandit_spec.rb (hunk 6 above)
Comment (ghbren) [reply spec/lib/salus/scanners/bandit_spec.rb:146]: no

**Comment by heimdall-asguard** [comment]
Review Error for cbtoni @ 2020-05-21 19:33:13 UTC
User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)

**Comment by heimdall-asguard** [comment]
Review Error for cbtoni @ 2020-05-21 19:42:03 UTC
User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)
